### PR TITLE
bpo-46677: Add examples of inheritance and attributes to `TypedDict` docs.

### PR DESCRIPTION
**Pull request by CharlieZhao95**
Add examples of inheritance and attributes to `TypedDict` docs.

https://bugs.python.org/issue46677

### CODE REVIEW

**Review by 97littleleaf11 — COMMENTED** [review]
I think it would be better to mention PEP655 in the doc!

**Review by JelleZijlstra — CHANGES_REQUESTED** [review]
Thanks! Some high-level feedback:

- It would be good to use the formal syntax for these attributes, as we do with ParamSpec.args (https://docs.python.org/3.10/library/typing.html#typing.ParamSpec.args). That way, we can refer to the attribute definitions by name and find them more easily.
- This is going to conflict with #31126. We should probably merge that one first as it got here first :)
- Could you also mention that the functional syntax is useful for TypedDicts with keys that aren't valid identifiers?
- We can't mention PEP 655 yet because the PEP is still in draft. In the meantime, we should mention the mixed inheritance trick to make a TypedDict with some required and non-required keys.

And cc @davidfstr in case you'd also like to give some feedback here.

**Comment by AlexWaygood** [comment]
Looks like something went haywire with a `git merge` here -- might be best to close this PR now and open a new one :)

**Comment by AlexWaygood** [comment]
> > Looks like something went haywire with a `git merge` here -- might be best to close this PR now and open a new one :)
> 
> I made a mistake about `git push`. Sincerely apologies to everyone. :(

It's alright -- happens to us all :) I've removed the requests for review that were autotriggered by Github's CODEOWNERS feature, and your force push has cleaned up the commit history -- so, no harm done :)

File: Doc/library/typing.rst [hunk 1]
```diff
@@ -1484,6 +1481,19 @@ These are not used in annotations. They are building blocks for declaring types.
       The keyword-argument syntax is deprecated in 3.11 and will be removed
       in 3.13. It may also be unsupported by static type checkers.
 
+   Furthermore, there is another reason to use the functional syntax. If the keys
+   of ``TypedDicts`` aren't valid identifiers, such as keywords, name with hyphens,
+   etc., we must use the functional syntax.
+   Example::
+
+      # raise SyntaxError
```
Comment (davidfstr) [inline Doc/library/typing.rst]: Nit: raise -> raises

File: Doc/library/typing.rst [hunk 2]
```diff
@@ -1484,6 +1481,19 @@ These are not used in annotations. They are building blocks for declaring types.
...
+      class Point2D(TypedDict):
+          in: int  # 'in'(a keyword)
```
Comment (davidfstr) [inline Doc/library/typing.rst]: Nit: Maybe `# 'in' is a keyword` would be more explicit?

File: Doc/library/typing.rst [hunk 3]
```diff
@@ -1500,6 +1510,70 @@ These are not used in annotations. They are building blocks for declaring types.
...
+         y: int
+         z: int
+
+   The type info for introspection can be accessed via ``Point2D.__annotations__``,
+   ``Point2D.__total__``, ``Point2D.__required_keys__``, and
+   ``Point2D.__optional_keys__``.
+
+   .. attribute:: __annotations__
+
+      ``Point2D.__annotations__`` gives a dict maps the field names to the field types::
```
Comment (davidfstr) [inline Doc/library/typing.rst]: Nit: `gives a dict maps` -> `gives a dict which maps`

File: Doc/library/typing.rst [hunk 4]
```diff
@@ -1500,6 +1510,70 @@ These are not used in annotations. They are building blocks for declaring types.
...
+         True
+         >>> class Point2D(TypedDict, total=False): pass
+         >>> Point2D.__total__
+         False
+
+   .. attribute:: __required_keys__
+   .. attribute:: __optional_keys__
+
+      ``Point2D.__required_keys__`` and ``Point2D.__optional_keys__`` return
+      ``frozenset`` objects containing required and optional keys, respectively.
```
Comment (davidfstr) [inline Doc/library/typing.rst]: PEP 655 (Required[] and NotRequired[] for TypedDict) deliberately uses the phrasing "potentially-missing keys" to refer to the concept described by `__optional_keys__` to avoid confusion with Optional[], which is an entirely different concept.

So:

* I might suggest replacing `optional keys` with `potentially-missing keys` here (and in similar locations).
* I could also see using the phrase `not-required keys` to have a different kind of alignment with the upcoming PEP 655.

FWIW, I think it's probably too late to change the name of `__optional_keys__` to avoid confusion with Optional[].

File: Doc/library/typing.rst [hunk 5]
```diff
@@ -1500,6 +1510,70 @@ These are not used in annotations. They are building blocks for declaring types.
...
+      Currently the only way to set required and optional keys is mixed inheritance,
+      declaring a ``TypedDict`` with one value for the ``total`` argument and
+      then inheriting it from another ``TypedDict`` with a different value for ``total``.
```
Comment (davidfstr) [inline Doc/library/typing.rst]: I'm looking forward to amending this with PEP 655 (Required[] and NotRequired[] for TypedDict). 😄 

(No changes suggested here.)

**Review by davidfstr — COMMENTED** [review]
Thanks for taking the time to update the documentation! It's a lot better with these changes, and removes the need to consult related PEPs for common scenarios.

I've left a few comments. Mostly nits.

**Comment by CharlieZhao95** [comment]
> I've left a few comments. Mostly nits.

Thanks for your patience! Your suggestions are definitely helpful to me. 😃

**Comment by davidfstr** [comment]
@CharlieZhao95 , a tiny nit on your [profile tagline](https://github.com/CharlieZhao95):

> **Sociable** programmer majoring in EE. 👻

Did amend spelling of first word.

(Normally I would send this type of comment to you directly as a Private Message (PM), but GitHub doesn't appear to support that.)

File: Doc/library/typing.rst [hunk 6]
```diff
@@ -1500,6 +1510,76 @@ These are not used in annotations. They are building blocks for declaring types.
...
    and makes all items defined in the class body required.
 
+   It is possible for a ``TypedDict`` type to inherit from one or more TypedDict types
+   using the class-based syntax.
+   Usage::
+
+      class Point3D(Point2D):
+         z: int
+
+   Now the ``Point3D`` has three items: ``x``, ``y`` and ``z``. It is equivalent to this
```
Comment (JelleZijlstra) [inline Doc/library/typing.rst]:
```suggestion
   ``Point3D`` has three items: ``x``, ``y`` and ``z``. It is equivalent to this
```

File: Doc/library/typing.rst [hunk 7]
```diff
@@ -1500,6 +1510,76 @@ These are not used in annotations. They are building blocks for declaring types.
...
+         ...     y: int
+         ...
+         >>> Point2D.__annotations__
+         {'x': <class 'int'>, 'y': <class 'int'>}
+
+   .. attribute:: __total__
+
+      ``Point2D.__total__`` gives the value of the ``total`` argument. Furthermore,
+      inherited items won't be affected by the totality flag which defined in superclass,
+      and instead use totality of the TypedDict type where they were defined.
```
Comment (JelleZijlstra) [inline Doc/library/typing.rst]: The second sentence is hard to parse, and feels out of place for the documentation of the `__total__` argument. I suggest to just omit it.

File: Doc/library/typing.rst [hunk 8]
```diff
@@ -1500,6 +1510,76 @@ These are not used in annotations. They are building blocks for declaring types.
...
+         False
+         >>> class Point3D(Point2D): pass
+         >>> Point3D.__total__
+         True
+
+   .. attribute:: __required_keys__
+   .. attribute:: __optional_keys__
+
+      ``Point2D.__required_keys__`` and ``Point2D.__optional_keys__`` return
+      ``frozenset`` objects containing required and potentially-missing keys, respectively.
```
Comment (JelleZijlstra) [inline Doc/library/typing.rst]:
```suggestion
      ``frozenset`` objects containing required and non-required keys, respectively.
```

File: Doc/library/typing.rst [hunk 9]
```diff
@@ -1500,6 +1510,76 @@ These are not used in annotations. They are building blocks for declaring types.
...
+      Currently the only way to set required and potentially-missing keys is mixed inheritance,
```
Comment (JelleZijlstra) [inline Doc/library/typing.rst]:
```suggestion
      Currently the only way to declare both required and non-required keys in the same ``TypedDict`` is mixed inheritance,
```

File: Doc/library/typing.rst [hunk 10]
```diff
@@ -1500,6 +1510,76 @@ These are not used in annotations. They are building blocks for declaring types.
    ``True`` as the value of the ``total`` argument. ``True`` is the default,
    and makes all items defined in the class body required.
 
+   It is possible for a ``TypedDict`` type to inherit from one or more TypedDict types
```
Comment (JelleZijlstra) [inline Doc/library/typing.rst]:
```suggestion
   It is possible for a ``TypedDict`` type to inherit from one or more other ``TypedDict`` types
```

File: Doc/library/typing.rst [hunk 11]
```diff
@@ -1484,6 +1481,19 @@ These are not used in annotations. They are building blocks for declaring types.
       The keyword-argument syntax is deprecated in 3.11 and will be removed
       in 3.13. It may also be unsupported by static type checkers.
 
+   Furthermore, there is another reason to use the functional syntax. If the keys
+   of ``TypedDicts`` aren't valid identifiers, such as keywords, name with hyphens,
+   etc., we must use the functional syntax.
```
Comment (JelleZijlstra) [inline Doc/library/typing.rst]:
```suggestion
   The functional syntax should also be used when any of the keys are not valid
   :ref:`identifiers`, for example because they are keywords or contain hyphens.
```
A bit shorter, and cross-link to https://docs.python.org/3.10/reference/lexical_analysis.html#identifiers where identifiers are defined.

File: Doc/library/typing.rst (hunk 10 above)
Comment (JelleZijlstra) [reply Doc/library/typing.rst]: We should probably also add that TypedDicts are not allowed to inherit from anything that is not a TypedDict, notably including `Generic`.

**Review by JelleZijlstra — CHANGES_REQUESTED** [review]
Thank you! Here are some comments.

**Comment by CharlieZhao95** [comment]
I have made the requested changes; please review again

File: Doc/library/typing.rst [hunk 12]
```diff
@@ -1500,6 +1509,93 @@ These are not used in annotations. They are building blocks for declaring types.
...
+
+   ``Point3D`` has three items: ``x``, ``y`` and ``z``. It is equivalent to this
+   definition::
+
+      class Point3D(TypedDict):
+          x: int
+          y: int
+          z: int
+
+   It is worth noting that a ``TypedDict`` cannot inherit from a non-TypedDict class,
```
Comment (JelleZijlstra) [inline Doc/library/typing.rst]:
```suggestion
   A ``TypedDict`` cannot inherit from a non-TypedDict class,
```

File: Doc/library/typing.rst [hunk 13]
```diff
@@ -1500,6 +1509,93 @@ These are not used in annotations. They are building blocks for declaring types.
...
+         False
+         >>> class Point3D(Point2D): pass
+         >>> Point3D.__total__
+         True
+
+   .. attribute:: __required_keys__
+   .. attribute:: __optional_keys__
+
+      ``Point2D.__required_keys__`` and ``Point2D.__optional_keys__`` return
+      ``frozenset`` objects containing required and non-required keys, respectively.
```
Comment (JelleZijlstra) [inline Doc/library/typing.rst]:
```suggestion
      :class:`frozenset` objects containing required and non-required keys, respectively.
```

**Review by JelleZijlstra — APPROVED** [review]

File: Doc/library/typing.rst [hunk 14]
```diff
@@ -1500,6 +1509,93 @@ These are not used in annotations. They are building blocks for declaring types.
...
+      ``Point2D.__annotations__`` gives a dict which maps the field names to the field types::
+
+         >>> from typing import TypedDict
+         >>> class Point2D(TypedDict):
+         ...     x: int
+         ...     y: int
+         ...
+         >>> Point2D.__annotations__
+         {'x': <class 'int'>, 'y': <class 'int'>}
+
```
Comment (Fidget-Spinner) [inline Doc/library/typing.rst]: I'm not in favour of this section. There's already an entry in [datamodel](https://docs.python.org/3/reference/datamodel.html) for `__annotations__` and and [even longer document](https://docs.python.org/3/howto/annotations.html#annotations-howto) by Larry. This feels like duplicate work.

File: Doc/library/typing.rst [hunk 15]
```diff
@@ -1500,6 +1509,93 @@ These are not used in annotations. They are building blocks for declaring types.
...
+      class Z(object): pass  # A non-TypedDict class
+
+      class XY(X, Y): pass  # OK
+
+      class XZ(X, Z): pass  # raises TypeError
+
+      T = TypeVar('T')
+      class XT(X, Generic[T]): pass  # raises TypeError
+
+   The type info for introspection can be accessed via ``Point2D.__annotations__``,
```
Comment (Fidget-Spinner) [inline Doc/library/typing.rst]: Maybe we can link the annotations howto here for `__annotations__` instead of asking the user to directly access the field.

**Review by Fidget-Spinner — COMMENTED** [review]
I just have one point. The rest LG. Thank you!

File: Doc/library/typing.rst [hunk 16]
```diff
@@ -1484,6 +1481,18 @@ These are not used in annotations. They are building blocks for declaring types.
       The keyword-argument syntax is deprecated in 3.11 and will be removed
       in 3.13. It may also be unsupported by static type checkers.
 
+   The functional syntax should also be used when any of the keys are not valid
+   :ref:`identifiers`, for example because they are keywords or contain hyphens.
```
Comment (kumaraditya303) [inline Doc/library/typing.rst:1485]:
```suggestion
   :ref:`identifiers`.
```

There can be other reasons and not a valid identifier is enough:
```py
from typing import TypedDict

A = TypedDict("A", {"a b c": int})
a = A()
a["a b c"] = 1
```

Comment (AlexWaygood) [reply Doc/library/typing.rst:1485]: The "for example" wording does not preclude the possibility that there are other reasons; it just gives examples of _some_ of the reasons

Comment (kumaraditya303) [reply Doc/library/typing.rst:1485]: Ya, but `not a valid identifier` is enough and no need to give examples as there is a link for identifiers given already.

Comment (AlexWaygood) [reply Doc/library/typing.rst:1485]: Hmm, I disagree — I don't think it's self-evident what the definition of an "identifier" is, and the link takes you (I think — haven't checked) to an extremely jargony page in the docs that's pretty scary tbh!

I think the examples here are helpful :)

Comment (kumaraditya303) [reply Doc/library/typing.rst:1485]: Okay I agree now

File: Doc/library/typing.rst (hunk 15 above)
Comment (CharlieZhao95) [reply Doc/library/typing.rst]: I agree this!
 Also, `__annotations__` is not a best practice in Python versions 3.10 and newer, perhaps it is not appropriate to be emphasized here. 

I will delete the section of `.. attribute:: __annotations__`, and suggest user to access `annotations-howto` directly.

**Review by AlexWaygood — APPROVED** [review]
This is a really great improvement — thank you!

**Review by kumaraditya303 — APPROVED** [review]

File: Doc/library/typing.rst [hunk 17]
```diff
@@ -1500,6 +1509,82 @@ These are not used in annotations. They are building blocks for declaring types.
...
+      class Z(object): pass  # A non-TypedDict class
+
+      class XY(X, Y): pass  # OK
+
+      class XZ(X, Z): pass  # raises TypeError
+
+      T = TypeVar('T')
+      class XT(X, Generic[T]): pass  # raises TypeError
+
+   The type info for introspection can be accessed via annotations dicts
```
Comment (JelleZijlstra) [inline Doc/library/typing.rst]:
```suggestion
   A ``TypedDict`` can be introspected via annotations dicts
```
"type info" sounds a bit too informal

**Review by JelleZijlstra — APPROVED** [review]
Thanks! @gvanrossum I'm planning to merge this docs enhancement for TypedDict.

**Review by Fidget-Spinner — APPROVED** [review]

**Comment by JelleZijlstra** [comment]
@CharlieZhao95 do you want to do the backports manually? Probably just need to fix some small conflicts.

**Comment by CharlieZhao95** [comment]
> 



> @CharlieZhao95 do you want to do the backports manually? Probably just need to fix some small conflicts.

Sure. It won't take long, I will finish it soon. 😉
